subsys/emds/emds.c

```diff
@@ -6,55 +6,353 @@
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		*size += ch->entry.len;
+		entries++;
+	}
+
+	return entries;
+}
+
 int emds_init(void)
 {
+	int rc;
+
+	if (emds_initialized) {
+		LOG_DBG("Already initialized");
+		return -EALREADY;
+	}
+
+	rc = emds_fs_init();
+	if (rc) {
+		return rc;
+	}
+
+	sys_slist_init(&emds_dynamic_entries);
+
+	emds_initialized = true;
+
+	LOG_DBG("Emergency Data Storage initialized.\n");
+
 	return 0;
 }
 
-int emds_entry_add(struct emds_dynamic_entry *entry)
+int emds_entry_add(struct emds_entry *entry)
 {
+	struct emds_dynamic_entry *ch;
+	struct emds_dynamic_entry *item;
+
+	if (!emds_initialized) {
+		return -ECANCELED;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		if (ch->entry.id == entry->id) {
+			return -EINVAL;
+		}
+	}
+
+	item = k_malloc(sizeof(struct emds_dynamic_entry));
+	if (!item) {
+		return -ENOSPC;
+	}
+
+	item->entry = *entry;
+	sys_slist_append(&emds_dynamic_entries, &item->node);
+
 	return 0;
 }
 
 int emds_store(void)
 {
+	if (!emds_ready) {
+		return -ECANCELED;
+	}
+
+	/* Lock all interrupts */
+	(void)irq_lock();
+	/* Start the emergency data storage process. */
+	k_sem_give(&emds_sem);
+
 	return 0;
 }
 
 int emds_load(void)
 {
+	struct emds_dynamic_entry *ch;
+
+	if (!emds_initialized) {
+		return -ECANCELED;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		ssize_t len = emds_flash_read(&default_emds_storage.cf_emds,
+					      ch->entry.id, ch->entry.data,
+					      ch->entry.len);
+
+		if (len == ch->entry.len) {
+			LOG_DBG("Read dynamic entry: %d, Data length: %d",
+				ch->entry.id, ch->entry.len);
+			LOG_HEXDUMP_DBG(ch->entry.data, ch->entry.len,
+					"Preserved data: ");
+		} else {
+			LOG_ERR("Read dynamic entry: (%d) failed (%d:%d).\n",
+				ch->entry.id, ch->entry.len, len);
+		}
+	}
+
+	STRUCT_SECTION_FOREACH(emds_entry, ch) {
+		ssize_t len = emds_flash_read(&default_emds_storage.cf_emds,
+					      ch->id, ch->data, ch->len);
+
+		if (len == ch->len) {
+			LOG_DBG("Read static entry: %d, Data length: %d",
+				ch->id, ch->len);
+			LOG_HEXDUMP_DBG(ch->data, ch->len, "Preserved data: ");
+		} else {
+			LOG_ERR("Read static entry: (%d) entry failed (%d:%d)",
+				ch->id, ch->len, len);
+		}
+	}
+
 	return 0;
 }
 
 int emds_clear(void)
 {
+	int rc;
+
+	if (!emds_initialized) {
+		return -ECANCELED;
+	}
+	
+	rc = emds_flash_clear(&default_emds_storage.cf_emds);
+	if (rc) {
+		return rc;
+	}
+
 	return 0;
 }
 
 int emds_prepare(void)
 {
+	int entries;
+	uint32_t size;
+	int rc;
+
+	if (!emds_initialized) {
+		return -ECANCELED;
+	}
+	
+	entries = emds_entries_size(&size);
+
+	rc = emds_flash_prepare(&default_emds_storage.cf_emds, entries, size);
+	if (rc) {
+		return rc;
+	}
+
+	emds_ready = true;
+
 	return 0;
 }
 
 uint32_t emds_store_time_get(void)
 {
-	return 0;
+	size_t block_size = default_emds_storage.cf_emds.flash_parameters->write_block_size;
+	uint32_t store_time_us;
```

Comment:
```suggestion
	uint32_t store_time_us = FLASH_BASE_TIME_OVERHEAD_US;
```
no need separate initialization